File: CountofRangeSum2.cpp
#include<iostream>
#include<vector>
#include<set>

using namespace std;

class Solution {
public:
    int countRangeSum(vector<int>& nums, int lower, int upper) {
    	
    	vector<long> sums(nums.size() + 1, 0);
    	
    	for(int i = 0; i < nums.size(); i++) {
    		sums[i + 1] = nums[i] + sums[i];
		}
		
		return countAndMergeSort(sums, 0, sums.size() - 1, lower, upper);
    	
    }
    
    int countAndMergeSort(vector<long>& sums, int begin, int end, int lower, int upper) {
    	
    	if(begin >= end) return 0;
    	
    	int mid = begin + (end - begin) / 2;
    	
    	int count = countAndMergeSort(sums, begin, mid, lower, upper) + countAndMergeSort(sums, mid + 1, end, lower, upper);
    	
    	vector<long> temp(end - begin + 1);
    	int l = mid + 1, j = mid + 1, k = mid + 1;
    	for(int i = begin, r = 0; i <= mid; i++, r++) {
    		while(j <= end && sums[j] - sums[i] < lower) j++;
    		while(k <= end && sums[k] - sums[i] <= upper) k++;
    		
    		count += k - j;
    		
			while(l <= end && sums[l] < sums[i]) temp[r++] = sums[l++];
			temp[r] = sums[i]; 
		}
		
		//copy(temp.begin(), temp.begin() + l - begin, sums.begin() + begin);
		
		for(int i = begin; i < l; i++) sums[i] = temp[i - begin];

		return count;
    	
	}
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	int lower, upper;
	cin>>lower>>upper;
	
	Solution *solution = new Solution();
	cout<<solution->countRangeSum(nums, lower, upper);
	
	return 0;
	
}

File: CountofRangeSum2.py
class Solution:
    def countRangeSum(self, nums, lower, upper):
        """
        :type nums: List[int]
        :type lower: int
        :type upper: int
        :rtype: int
        """
        def findIndex(sums, val):
            l, r = 0, len(sums) - 1
            while l <= r:
                mid = l + (r - l) //2
                if sums[mid] > val:
                    r = mid - 1
                else:
                    l = mid + 1
            return l

        def countRangeSum(begin, end):
            if begin == end:
                return 1 if lower <= nums[begin] <= upper else 0
            if begin > end:
                return 0
            mid = begin + (end - begin) // 2
            s = 0
            sums = []
            for i in range(mid + 1, end + 1):
                s += nums[i]
                sums.append(s)
            sums.sort()

            s = 0
            count = 0
            for i in range(mid, begin - 1, -1):
                s += nums[i]
                count += findIndex(sums, upper - s + 0.5) - findIndex(sums, lower - s - 0.5)

            return count + countRangeSum(begin, mid) + countRangeSum(mid + 1, end)

        return countRangeSum(0, len(nums) - 1)


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    lower = int(input())
    upper = int(input())
    solution = Solution()
    print(solution.countRangeSum(nums, lower, upper))
